target/linux/realtek/dts-5.10/rtl839x.dtsi

```diff
@@ -0,0 +1,234 @@
+		reg = <##n>; \
+		compatible = "ethernet-phy-ieee802.3-c22"; \
+	};
+
+#define EXTERNAL_SFP_PHY(n) \
+	phy##n: ethernet-phy@##n { \
+		compatible = "ethernet-phy-ieee802.3-c22"; \
+		sfp; \
+		media = "fibre"; \
+		reg = <##n>; \
+	};
+
+#define SWITCH_PORT(n, s, m) \
+	port@##n { \
+		reg = <##n>; \
+		label = SWITCH_PORT_LABEL(s) ; \
+		phy-handle = <&phy##n>; \
+		phy-mode = #m ; \
+	};
+
+#define SWITCH_SFP_PORT(n, s, m) \
+	port@##n { \
+		reg = <##n>; \
+		label = SWITCH_PORT_LABEL(s) ; \
+		phy-handle = <&phy##n>; \
+		phy-mode = #m ; \
+		fixed-link { \
+			speed = <1000>; \
+			full-duplex; \
+		}; \
+	};
+
+/ {
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	compatible = "realtek,rtl838x-soc";
+
+	cpus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		frequency = <700000000>;
+
+		cpu@0 {
+			compatible = "mips,mips34Kc";
+			reg = <0>;
+		};
+	};
+
+	memory@0 {
+		device_type = "memory";
+		reg = <0x0 0x8000000>;
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,38400";
+	};
+
+	lx_clk: lx_clk {
+		compatible = "fixed-clock";
+		#clock-cells = <0>;
+		clock-frequency = <200000000>;
+	};
+
+	cpuintc: cpuintc {
+		compatible = "mti,cpu-interrupt-controller";
+		#address-cells = <0>;
+		#interrupt-cells = <1>;
+		interrupt-controller;
+	};
+
+	soc: soc {
+		compatible = "simple-bus";
+		#address-cells = <1>;
+		#size-cells = <1>;
+		ranges = <0x0 0x18000000 0x10000>;
+
+		intc: interrupt-controller@3000 {
+			compatible = "realtek,rtl8380-intc", "realtek,rtl-intc";
+			reg = <0x3000 0x18>, <0x3018 0x18>;
+			interrupt-controller;
+			#interrupt-cells = <2>;
+
+			interrupt-parent = <&cpuintc>;
+			interrupts = <2>, <3>, <4>, <5>, <6>;
+		};
+
+		spi0: spi@1200 {
+			compatible = "realtek,rtl8380-spi";
+			reg = <0x1200 0x100>;
+
+			#address-cells = <1>;
+			#size-cells = <0>;
+		};
+
+		uart0: uart@2000 {
+			compatible = "ns16550a";
+			reg = <0x2000 0x100>;
+
+			clocks = <&lx_clk>;
+
+			interrupt-parent = <&intc>;
+			interrupts = <31 1>;
+
+			reg-io-width = <1>;
+			reg-shift = <2>;
+			fifo-size = <1>;
+			no-loopback-test;
+		};
+
+		uart1: uart@2100 {
+			pinctrl-names = "default";
+			pinctrl-0 = <&enable_uart1>;
+
+			compatible = "ns16550a";
+			reg = <0x2100 0x100>;
+
+			clocks = <&lx_clk>;
+
+			interrupt-parent = <&intc>;
+			interrupts = <30 2>;
+
+			reg-io-width = <1>;
+			reg-shift = <2>;
+			fifo-size = <1>;
+			no-loopback-test;
+
+			status = "disabled";
+		};
+
+		watchdog0: watchdog@3150 {
+			compatible = "realtek,rtl8390-wdt";
+			reg = <0x3150 0xc>;
+
+			realtek,reset-mode = "soc";
+
+			clocks = <&lx_clk>;
+			timeout-sec = <30>;
+
+			interrupt-parent = <&intc>;
+			interrupt-names = "phase1", "phase2";
+			interrupts = <19 2>, <18 2>;
+		};
+
+	};
+
+	gpio0: gpio-controller@b8003500 {
+		compatible = "realtek,rtl838x-gpio";
+		reg = <0xb8003500 0x20>;
+		gpio-controller;
+		#gpio-cells = <2>;
+		interrupt-parent = <&intc>;
+		interrupts = <23 2>;
+	};
+
+	gpio1: rtl8231-gpio {
```

Comment:
I would rather see the bindings for these GPIO expanders in the device-specific DTS. The RTL8231 isn't part of the SoC, so it depends on the device whether or not this is present and should be specified.

Reply:
We have done it like this for all other SoC architectures. I also still have to find an RTL device without that chip. It is also good to have the right SMI bus number in here.

Reply:
But there is no sane default for "the right SMI bus number". The Panasonic switches that @musashino205 is working on don't have the RTL8231 as a GPIO expander, but different chips. There a delete statement is required to get rid of the rtl8231 definition.

Reply:
To add to my previous comment, "we have always done it like this" neglects the fact that insights evolve over time, and old code can be amended. Having the board-specific `indirect-access-bus-id = <3>` in a DT include for the SoC definition doesn't make any sense to me, so I stand by my point that this doesn't belong in this file.

Reply:
Moved.

Reply:
Yes, I agree, this should be a proper MDIO device, ie. registered in the MDIO bus. Introducing a new DTS attribute `indirect-access-bus-id` (even without a vendor prefix...) is not the way.

Reply:
I have tried to upstream a driver, to implement this as a proper MDIO device, but it got stuck on some regmap changes.
See: https://lore.kernel.org/all/cover.1623532208.git.sander@svanheule.net/

We could include those as pending/hack patches, until I have time to properly resolve this with upstream. Note that for now this would require a bitbanged MDIO bus, as there is no support for the auxiliary MDIO peripheral (as an MDIO bus) yet.

Reply:
The RTL8231 is not always connected via MDIO to the SoC. It has 3 different interface mode according to its specs: SMI, MDIO and Serial (and yes, they make a difference between SMI and MDIO). At this point we do not define different SMI/MDIO/ShiftRegister busses. One reason is that this is completely transparent on the RTL839x platform. For the XGS1250 we have the RTL8231 in shift register mode. Which seems to be set up already by u-boot. But maybe we want to do this from the .dts, too.

Reply:
* Serial mode: no definition in DT required, since this is only used by the hardware port LED peripheral. The kernel/hardware interface in this case is the LED peripheral. In this mode, no internal registers of the RTL8231 can be accessed and it behaves as a plain 36 (or was it 37?) bit shift register.
* MDIO mode: This is the mode all known implementation use, if they require GPIOs from the RTL8231.
* SMI mode: Actually an I2C mode. Already supported by my driver, thanks to the regmap implementation.

These modes cannot be runtime selected, and are determined by the bootstrapping pins of the RTL8231. If the XGS1250 uses an RTL8231 in shift register mode as a GPIO expander, then this is already supported by the kernel as a generic shift register with 36 (37) bits.

Reply:
Please no bit-banged MDIO to steer LEDs if the existing driver supports hardware accelerated access and is widely tested. It can be easily changed to look like an MDIO device if necessary. Also I wonder why complicate things in the .dts if it can anyway not be configured there...

Reply:
I'm also not a fan of bitbanged MDIO, but an driver for the auxiliary MDIO controller should be quite easy to implement (I may have some patches somewhere still...).

The devicetree description only follows the hardware description, we're not making more complicated than required. The SoC has an MDIO bus master, and that MDIO bus has the RTL8231 as one of the child devices. And that's how it would be described in the DT.

Reply:
That sounds good. But not now, please. This would need to be tested on several of the supported devices first, as this basically concerns them all.

Reply:
The AUX MDIO looks very straight forward and simple to use according to the register defintions. I'm sure I can write a driver for it from scratch in less than one evening of work. But yes, let's postpone this for now and not make things more complicated for this PR.

Reply:
@dangowrt Found my old code: https://github.com/svanheule/linux/commit/d91873413dc3caed4999f7c178ff93e29c6d39dd